onnxruntime/core/graph/graph.cc

```diff
@@ -316,23 +316,31 @@ void Node::SetExecutionProviderType(ProviderType execution_provider_type) {
   execution_provider_type_ = execution_provider_type;
 }
 
-void Node::ToProto(NodeProto& proto) const {
-  // Set name.
+void Node::ToProto(NodeProto& proto, bool update_subgraphs) const {
   proto.set_name(name_);
-  // Set op type.
   proto.set_op_type(op_type_);
-  // Set op domain;
-  proto.set_domain(domain_);
-  // Set doc string.
-  proto.set_doc_string(description_);
+
+  if (!domain_.empty())
+    proto.set_domain(domain_);
+
+  if (!description_.empty())
+    proto.set_doc_string(description_);
 
   // Set attributes.
   proto.clear_attribute();
   for (const auto& attribute : attributes_) {
     const gsl::not_null<AttributeProto*> attr{proto.add_attribute()};
     *attr = attribute.second;
-  }
 
+    if (update_subgraphs && attribute.second.has_g()) {
+      // partitioning or optimization may have changed the subgraph, so the original GraphProto may not be valid.
+      // call ToGraphProto to get the current info and use this for the attribute
+      const auto& latest = attr_to_subgraph_map_.find(attribute.first)->second->ToGraphProto();
```

Comment:
Can we know when the subgraph proto needs re-building, so that we avoid it otherwise?

Reply:
ToGraphProto only rebuilds if it has to.